solidity/contracts/liquidity-protection/LiquidityProtectionStats.sol

```diff
@@ -0,0 +1,231 @@
+// SPDX-License-Identifier: SEE LICENSE IN LICENSE
+pragma solidity 0.6.12;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/utils/EnumerableSet.sol";
+
+import "./interfaces/ILiquidityProtectionStats.sol";
+import "../utility/Utils.sol";
+import "../utility/SafeMath.sol";
+import "../token/interfaces/IDSToken.sol";
+import "../token/interfaces/IERC20Token.sol";
+
+/**
+ * @dev This contract aggregates the statistics of the liquidity protection mechanism.
+ */
+contract LiquidityProtectionStats is ILiquidityProtectionStats, AccessControl, Utils {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using SafeMath for uint256;
+
+    bytes32 public constant ROLE_SUPERVISOR = keccak256("ROLE_SUPERVISOR");
+    bytes32 public constant ROLE_SEEDER = keccak256("ROLE_SEEDER");
+    bytes32 public constant ROLE_OWNER = keccak256("ROLE_OWNER");
+
+    mapping(IDSToken => uint256) private _totalPoolAmount;
+    mapping(IDSToken => mapping(IERC20Token => uint256)) private _totalReserveAmount;
+    mapping(IDSToken => mapping(IERC20Token => mapping(address => uint256))) private _totalProviderAmount;
+
+    mapping(address => EnumerableSet.AddressSet) private _providerPools;
+
+    // allows execution only by an owner
+    modifier ownerOnly {
+        require(hasRole(ROLE_OWNER, msg.sender), "ERR_ACCESS_DENIED");
+        _;
+    }
+
+    // allows execution only by a seeder
+    modifier seederOnly {
+        require(hasRole(ROLE_SEEDER, msg.sender), "ERR_ACCESS_DENIED");
+        _;
+    }
+
+    constructor() public {
+        // set up administrative roles
+        _setRoleAdmin(ROLE_SUPERVISOR, ROLE_SUPERVISOR);
+        _setRoleAdmin(ROLE_SEEDER, ROLE_SUPERVISOR);
+        _setRoleAdmin(ROLE_OWNER, ROLE_SUPERVISOR);
+
+        // allow the deployer to initially govern the contract
+        _setupRole(ROLE_SUPERVISOR, msg.sender);
+    }
+
+    /**
+     * @dev increases the total amounts
+     * can only be executed only by an owner
+     *
+     * @param provider          liquidity provider address
+     * @param poolToken         pool token address
+     * @param reserveToken      reserve token address
+     * @param poolAmount        pool token amount
+     * @param reserveAmount     reserve token amount
+     */
+    function increaseTotalAmounts(
+        address provider,
+        IDSToken poolToken,
+        IERC20Token reserveToken,
+        uint256 poolAmount,
+        uint256 reserveAmount
+    ) external override ownerOnly {
+        _totalPoolAmount[poolToken] = _totalPoolAmount[poolToken].add(poolAmount);
+        _totalReserveAmount[poolToken][reserveToken] = _totalReserveAmount[poolToken][reserveToken].add(reserveAmount);
+        _totalProviderAmount[poolToken][reserveToken][provider] = _totalProviderAmount[poolToken][reserveToken][
+            provider
```

Comment:
formatting: something is off here - is that prettier?

Reply:
Yes, it is (I've mentioned that in the past BTW, it is also the case in many other places in the code).

Reply:
ok cool